frontends/ast/simplify.cc

```diff
@@ -923,6 +923,19 @@ bool AstNode::simplify(bool const_fold, bool at_zero, bool in_lvalue, int stage,
 	bool children_are_self_determined = false;
 	bool reset_width_after_children = false;
 
+	if ((type == AST_GENFOR || type == AST_FOR) && children.size() != 0)
+	{
+		AstNode *init_ast = children[0];
+		if (init_ast->type == AST_ASSIGN_EQ && init_ast->children[0]->type != AST_IDENTIFIER) {
+			auto *clone = init_ast->children[0]->clone();
+			clone->simplify(false, false, true, stage, -1, false, in_param);
+			current_ast_mod->children.push_back(clone);
+			init_ast->children[0]->type = AST_IDENTIFIER;
+			init_ast->children[0]->id2ast = clone;
+			init_ast->children[0]->children.clear();
+		}
+	}
```

Comment:
This strategy doesn't quite for for something like:
```SystemVerilog
module top(out);
    output integer out;
    initial begin
        int i;
        i = 10;
        for (int i = 0; i < 5; i++)
            out = i;
        out = i;
    end
endmodule
```
Other tools seem to workaround this issue _for procedural loops_ by wrapping them in a synthetic named block. However, this of cannot be done for the generate loops, as adding a synthetic scope would affect the elaborated naming.

You may also be able to use some `current_scope`/`backup_scope` shenanigans to workaround this issue.

Reply:
I added synthetic named block just to procedural loops and added above example to the tests.

Reply:
There is an analogous issue for generate loops:
```SystemVerilog
`default_nettype none
module top(out);
    output integer out;
    logic x;
    for (genvar x = 0; x < 2; x++) begin : loop
        logic [x:0] y;
    end
    assign x = 1'sb1;
    assign loop[0].y = 1'sb1;
    assign loop[1].y = 1'sb1;
endmodule
```
Note that simply wrapping it in another block doesn't work here as this would affect the elaborated naming of `loop` and its contents.

Reply:
I fixed this issue by renaming in genvar declared in loop declaration (and all descendant usages of this variable). This fixes above problem without changing elaborated naming of ``loop``.